Memory improvements

amoerie opened the pull request:
Fixes #1262

#### Checklist
- [x] The pull request branch is in sync with latest commit on the *fo-dicom/development* branch
- [x] I have updated API documentation
- [x] I have included unit tests
- [ ] I have updated the change log
- [x] I am listed in the CONTRIBUTORS file

#### Changes proposed in this pull request:
- Reduce memory allocations by reusing buffers, either internally or from a shared arraypool

#### To Do

- [ ] Migrate usages of IByteBuffer.GetByteRange(int offset, int count) to IByteBuffer.GetByteRange(int offset, int count, byte[] output) where possible
- [ ] Write PDU directly to network stream (if possible)

amoerie committed:
Add class to determine free ports for the benchmarks

amoerie committed:
Move NopCStoreProvider there too

amoerie committed:
Disable logging in server benchmarks, it only slows things down

amoerie committed:
Expose NullLoggerManager, no harm in that I think

amoerie committed:
Update server benchmarks to avoid socket errors

amoerie committed:
Don't allocate a new byte array for the preamble in a RawPDU

amoerie committed:
Aggressively inline the offset check when reading a raw PDU

amoerie committed:
Use a buffer from the arraypool to read incoming PDU data

amoerie committed:
Remove [Obsolete] attribute from IByteBuffer.GetByteRange, there are still valid use cases

amoerie committed:
Delete IByteBuffer.GetByteRange and replace with usages of IByteBuffer.GetByteRange that takes an output byte array

amoerie committed:
Add missing implementation of EvenLengthBuffer.CopyToStream[Async]

amoerie committed:
Fix broken test

amoerie committed:
Mark tests as a library project so it's not runnable independently

amoerie committed:
Improve error stack traces in Dicom Network

More specifically, use ExceptionDispatchInfo.Capture to restore the original exception state and rethrow the original exception then. This considerably improves stack traces and traceability for errors.

amoerie committed:
Configure whether the BinaryReader/Writer classes should leave the inner stream open or not when disposed

amoerie committed:
Use a dedicated BufferedStream for writing to the network

amoerie committed:
Stream PDATA-TF to the network instead of materializing it in memory first

gofal commented on FO-DICOM.Core/IO/Buffer/EvenLengthBuffer.cs:
```diff
@@ -69,32 +69,16 @@ public void GetByteRange(long offset, int count, byte[] output)
-                }
-            }
+            s.Write(Buffer.Data, 0, (int)Buffer.Size);
+            s.WriteByte(0);
```
Is that really the same behavior? previously if count was higher than the length of the buffer, then the rest is filled with zeroes. Now there is always exactly one Zero written, without any check or conditional. Is that intended?

amoerie replied:
Yes. 
You only make an even length buffer when the buffer is uneven. Look at the Size property. That is why we just need to write the full buffer and add one zero.

Previously, you could pass in a offset and length property. If the length was (much) bigger, the stream would be padded with zeroes to ensure length bytes were written. But in reality, we never used this offset and length parameter, we always write the PDU in full to the network stream.

gofal replied:
I see. While the old code was kind of self-explaining what it does, the new code here is optimized but not intuitive any more. This should then be explained with a comment so so.

amoerie replied:
I added code comments in the new PR

amoerie committed:
Reduce the amortized memory usage of PDU writing to almost zero

amoerie committed:
When parsing incoming DICOM messages, use rented byte arrays if possible

Furthermore, immediately construct the MemoryStream with the correct capacity to avoid internal doubling of the capacity (and incurring a memory allocation + Buffer.BlockCopy penalty after every doubling)

amoerie committed:
Allow read/write from PDU with buffer, index and count

amoerie committed:
Ensure 0 in second byte of preamble

When lending an array from the pool, you are not guaranteed to get a zeroed one

amoerie committed:
Use rented byte arrays when reading raw PDUs

This required making PDataTF and PDV disposable, so that their respective rented arrays can be returned when the objects are disposed

amoerie commented:
Note: I'm focusing on build / v5-net-core-31 for the moment. I'll figure out why the tests are failing in .NET Framework later.
Since I'm not making any changes in the v4 codebase, I'm a bit puzzled how stuff is failing there too.

amoerie commented:
I'm a bit puzzled why that test is failing. This is the stack trace:

```
[xUnit.net 00:00:28.94]     FellowOakDicom.Tests.IO.Buffer.TempFileBufferTest.GetByteRange_CompareWithInitializer_ExactMatch(offset: 0, count: 128) [FAIL]
  Failed FellowOakDicom.Tests.IO.Buffer.TempFileBufferTest.GetByteRange_CompareWithInitializer_ExactMatch(offset: 0, count: 128) [7 ms]
  Error Message:
   System.IO.FileNotFoundException : Could not find file 'C:\Users\runneradmin\AppData\Local\Temp\tmp6ACC.tmp'.
  Stack Trace:
     at System.IO.__Error.WinIOError(Int32 errorCode, String maybeFullPath)
   at System.IO.FileStream.Init(String path, FileMode mode, FileAccess access, Int32 rights, Boolean useRights, FileShare share, Int32 bufferSize, FileOptions options, SECURITY_ATTRIBUTES secAttrs, String msgPath, Boolean bFromProxy, Boolean useLongPath, Boolean checkHost)
   at System.IO.FileStream..ctor(String path, FileMode mode, FileAccess access, FileShare share)
   at FellowOakDicom.IO.FileReference.OpenRead() in D:\a\fo-dicom\fo-dicom\FO-DICOM.Core\IO\FileReference.cs:line 101
   at FellowOakDicom.IO.Buffer.TempFileBuffer.GetByteRange(Int64 offset, Int32 count, Byte[] output) in D:\a\fo-dicom\fo-dicom\FO-DICOM.Core\IO\Buffer\TempFileBuffer.cs:line 81
   at FellowOakDicom.Tests.IO.Buffer.TempFileBufferTest.GetByteRange_CompareWithInitializer_ExactMatch(Int32 offset, Int32 count) in D:\a\fo-dicom\fo-dicom\Tests\FO-DICOM.Tests\IO\Buffer\TempFileBufferTest.cs:line 41
```

And this is the test:

```
        [Theory]
        [InlineData(20, 150)]
        [InlineData(0, 128)]
        [InlineData(5, 1)]
        [InlineData(30, 224)]
        [InlineData(12, 0)]
        public void GetByteRange_CompareWithInitializer_ExactMatch(int offset, int count)
        {
            var data = Enumerable.Range(0, 254).Select(i => (byte)i).ToArray();
            var buffer = new TempFileBuffer(data);

            var expected = new ArraySegment<byte>(data, offset, count);
            var actual = new byte[count];
            buffer.GetByteRange(offset, count, actual);
            Assert.Equal(expected, actual);
        }
```

For some reason, the temporary file is removed almost immediately after creating it, and the GetByteRange call fails because the temporary file is gone. Not sure what my changes have to do with this, but it sure looks like it started breaking because of me.

Will investigate later.

amoerie committed:
Ensure that benchmarks do not exhaust client side socket ports

amoerie committed:
Fix typo

amoerie committed:
Do not expose PDU.GetCommonFields to the outside world

amoerie committed:
Remove StreamByteBuffer.GetByteRange, we never use it anymore

amoerie committed:
Use rented byte arrays when sending PDUs

This modifies the PDataTFStream. Instead of allocating new byte arrays and filling them, we now rent byte arrays from the pool and pass them to the PDU once they're all filled up. Thanks to the disposal of PDUs, they get returned to the pool in the end.

amoerie committed:
Add safety nets around PDU and PDV so that the internal rented arrays are always disposed

amoerie commented:
@gofal  and @mrbean-bremen this is ready for a review. I still need to figure out why the temp file related test is failing in .NET Framework, but as you can see all tests are green on the other platforms. So it would be constructive to get your opinions already.

I might have more time later this week, but nothing is guaranteed. Thanks in advance for the feedback!

amoerie committed:
Fix bug in StreamByteBuffer where too much data is written

amoerie committed:
Only use allowed length of rented buffer when writing from FileByteBuffer

amoerie committed:
Prevent integer overflow when reading from a large file/stream > 2GB

amoerie committed:
Add messages system to TempFile system to debug mysterious deleted temp file in .NET framework

Re-add default ctor for TempFileBuffer, this is important

amoerie committed:
Add messages system to TempFile system to debug mysterious deleted temp file in .NET framework

amoerie committed:
Re-add default ctor for TempFileBuffer, this is important

amoerie committed:
Revert "Re-add default ctor for TempFileBuffer, this is important"

This reverts commit 6f5639bacfb77f81a26c0aa1b51b901b002b9214.

amoerie committed:
Revert "Add messages system to TempFile system to debug mysterious deleted temp file in .NET framework"

This reverts commit 23fe6ea198913585a2c1ad189408501c16ff9fed.

amoerie commented:
Ugh. When I add some diagnostic tools to investigate the temp file issue, the problem disappears. I've reverted my diagnostic changes, let's see if the issue reappears now...

amoerie committed:
Reinstate all benchmarks

amoerie commented:
Closing this pull request, will be split up into multiple smaller parts.

gofal commented on FO-DICOM.Core/IO/Buffer/TempFileBuffer.cs:
```diff
@@ -13,6 +14,8 @@ namespace FellowOakDicom.IO.Buffer
     /// </summary>
     public sealed class TempFileBuffer : IByteBuffer
     {
+        private readonly List<string> _messages;
```
here messages are collected. Where are they evaluated?

amoerie replied:
That was a temporary addition to diagnose the temp file test that is failing. These changes are not intended to be merged.
I did this because I the tests are always green on my machine, so I need to add diagnostics that are visible in the CI output.
You can ignore this, I will not re-add it again.
